servant-server/src/Servant/Server/Internal/RoutingApplication.hs

```diff
@@ -264,18 +303,17 @@ passToServer Delayed{..} x =
 runDelayed :: Delayed env a
            -> env
            -> Request
-           -> CleanupRef
-           -> IO (RouteResult a)
-runDelayed Delayed{..} env req cleanupRef =
+           -> ResourceT IO (RouteResult a)
+runDelayed Delayed{..} env req  =
   runDelayedIO
     (do c <- capturesD env
         methodD
         a <- authD
         b <- bodyD
-        DelayedIO $ \ r _cleanup -> return (serverD c a b r)
+        r <- ask
+        returnRouteResult (serverD c a b r)
```

Comment:
Now that `Handler` is abstract, and that this is being added, I feel like it makes sense to make `Handler` a `MonadResource` so that handlers can also register, allocate, and release. That said, that can just as well be a separate PR.